### PR TITLE
Fix for ODR-931.  Instead of getting a  range of sel entries, the sel…

### PR DESCRIPTION
**Pull request by tannoa2**
… poller alerts now traverse the sel log using a recursive

### CODE REVIEW

**Comment by rackhdmirror** [comment]
@RackHD/corecommitters please review.

**Comment by tannoa2** [comment]
Test this please

**Comment by tannoa2** [comment]
@jlongever @uppalk1 @zyoung51 @keedya @geoff-reid

File: lib/jobs/ipmi-job.js [hunk 1]
```diff
@@ -232,91 +297,42 @@ function ipmiJobFactory(
...
-                decodedSel["Event Type Code"] = eventType;
-                decodedSel["Sensor Type Code"] = rawSelRecord[10];
-                parsedSel[index] = _.merge(parsedSel[index], decodedSel);
+            lastPreviouslyReadSelEntry = _.get(workObj,'config.lastPreviouslyReadSelEntry', "-1");
+
+            //counter will keep track of the number of recursions and it should be < totalEntries
+            var counter = 0;
+            var id="0x"+(parseInt(lastPreviouslyReadSelEntry,16)+1).toString(16);
+            return self.loop(selArray, id , data, counter, totalEntries)
+            .then(function(resutls) {
```
Comment (anhou) [inline lib/jobs/ipmi-job.js]: typo. 'resutls'

File: lib/jobs/ipmi-job.js [hunk 2]
```diff
@@ -223,6 +223,71 @@ function ipmiJobFactory(
...
+     *
+     * @param selArray
+     * @param selRecordID
+     * @param data
+     * @param counter
+     * @param totalEntries
+     */
+    IpmiJob.prototype.loop = function(selArray, selRecordID, data, counter, totalEntries) {
+        var self = this;
+        var dataCloned = (JSON.parse(JSON.stringify(data)));//cloning the data object
```
Comment (anhou) [inline lib/jobs/ipmi-job.js]: no need for outside bracket

File: lib/jobs/ipmi-job.js [hunk 3]
```diff
@@ -223,6 +224,74 @@ function ipmiJobFactory(
...
+                    .then(function(rawRead){
+                        // we add two properties ("Sensor Type Code" and "Event Type Code") to each sel entry.
+                        var decodedSel = {
+                            "Event Type Code" : "",
+                            "Sensor Type Code" : ""
+                        };
+                        var regex = new RegExp('\r?\n','g');
+                        var rawRead = rawRead.replace(regex, '');
+                        var values = rawRead.split(" ");
+                        var mask = 1 << 7; // a mask to clear the 7th bit
```
Comment (anhou) [inline lib/jobs/ipmi-job.js]: @tannoa2  you could add '/*jslint bitwise: true */' to silence the warnings

File: lib/jobs/ipmi-job.js [hunk 4]
```diff
@@ -223,6 +223,71 @@ function ipmiJobFactory(
...
+                        values = values.slice(1,19);// remove empty entry
+                        var rawSelRecord = values.slice(2,18);
+                        var nextRecordID = values.slice(0,2);//This is the next SEL entry
+                        //This is from table 32 of the IPMI spec v4
+                        var pad = "00";
+                        var eventType = (parseInt(rawSelRecord[12],16) & ~mask).toString(16).slice(-2);
```
Comment (anhou) [inline lib/jobs/ipmi-job.js]: @tannoa2 you could add '/*jslint bitwise: true */' to silence the warnings '&' ad '~'

File: lib/jobs/ipmi-job.js [hunk 5]
```diff
@@ -223,6 +223,71 @@ function ipmiJobFactory(
...
+                        eventType = pad.substring(0, pad.length - eventType.length) + eventType;
+                        decodedSel["Event Type Code"] = eventType;
+                        decodedSel["Sensor Type Code"] = rawSelRecord[10];
+                        parsedEntry[0] = _.merge(parsedEntry[0], decodedSel);
+                        selArray.push(parsedEntry[0]);
+
+                        //if  the next record id is "ff ff"
+                        //it mean we reached the end of the sel log
+                        if( nextRecordID[0] !== "ff" && nextRecordID[1] !== "ff" && counter < totalEntries){
+                            return Promise.try(function() {
```
Comment (brianparry) [inline lib/jobs/ipmi-job.js]: @tannoa2 Shouldn't you be checking counter before calling `getSelEntry`?

**Comment by geoff-reid** [comment]
:+1:

**Pull request merged**